click/_compat.py

```diff
@@ -512,7 +519,19 @@ def open_stream(filename, mode='r', encoding=None, errors='strict',
     else:
         f = os.fdopen(fd, mode)
 
-    return _AtomicFile(f, tmp_filename, os.path.realpath(filename)), True
+    real_filename = os.path.realpath(filename)
+
+    # Get permissions to set on the target file. If the target file already
+    # exists, retain its current permissions. If the target file will be
+    # created, respect the permissions from the current umask.
+    permissions = None
+    if not WIN:
+        try:
+            permissions = stat.S_IMODE(os.stat(real_filename).st_mode)
+        except OSError:
+            permissions = 0o666 & ~_get_current_umask()
```

Comment:
IMO for the fallback it's better to set here a default permission, like 0o666 or even more paranoidal 0o600. As @andersk right said it's not thread-safe.

Reply:
Or even better, just `permissions = None`.

Reply:
The default permissions if you didn’t specify `atomic=True` are implicitly `0o666 & ~umask`, because that’s the default behavior of `open()` (no explicit umask computation is required). So it’s unexpected for `atomic=True` to result in different permissions. `permissions = 0o666` is definitely too permissive, as `chmod` is *not* implicitly masked by `umask`, while `permissions = 0o600` (or `pass`, which has the same effect) is too restrictive.

The underlying problem is that `tempfile` simply has the wrong permissions behavior for this use case. It’s designed for temporary files that will later be deleted, not for creating files with which to atomically replace other files.

Reply:
You are right, anyways at least on save an existing file there will be correct permissions.